wandb/integration/xgboost/xgboost.py

```diff
@@ -0,0 +1,181 @@
+        self.log_feature_importance = log_feature_importance
+        self.importance_type = importance_type
+        self.define_metric = define_metric
+
+    def before_training(self, model):
+        """Run before training is finished"""
+
+        # Update W&B config
+        config = model.save_config()
+        wandb.config.update(json.loads(config))
+
+        return model
+
+    def after_training(self, model):
+        """Run after training is finished."""
+
+        # Log the booster model as artifacts
+        if self.log_model:
+            self._log_model_as_artifact(model)
+
+        # Plot feature importance
+        if self.log_feature_importance:
+            self._log_feature_importance(model)
+
+        # Log the best score and best iteration
+        if model.attr("best_score") is not None:
+            wandb.log(
+                {
+                    "best_score": float(cast(str, model.attr("best_score"))),
+                    "best_iteration": int(cast(str, model.attr("best_iteration"))),
+                }
+            )
+
+        return model
+
+    def after_iteration(self, model, epoch, evals_log):
+        """Run after each iteration. Return True when training should stop."""
+
+        # Log metrics
+        for data, metric in evals_log.items():
+            for metric_name, log in metric.items():
+                if self.define_metric:
+                    self._define_metric(data, metric_name)
+                    wandb.log({f"{data}-{metric_name}": log[-1]}, commit=False)
+                else:
+                    wandb.log({f"{data}-{metric_name}": log[-1]}, commit=False)
+
+        wandb.log({"epoch": epoch})
+
+        self.define_metric = False
+
+        return False
+
+    def _log_model_as_artifact(self, model):
+        model_name = f"{wandb.run.id}_model.json"
+        model_path = f"{wandb.run.dir}/{model_name}"
+        model.save_model(str(model_path))
+
+        model_artifact = wandb.Artifact(name=model_name, type="model")
+        model_artifact.add_file(model_path)
+        wandb.log_artifact(model_artifact)
+
+    def _log_feature_importance(self, model):
+        fi = model.get_score(importance_type=self.importance_type)
+        fi_data = [[k, fi[k]] for k in fi]
+        table = wandb.Table(data=fi_data, columns=["Feature", "Importance"])
+        wandb.log(
+            {
+                "Feature Importance": wandb.plot.bar(
+                    table, "Feature", "Importance", title="Feature Importance"
+                )
+            }
+        )
+
+    def _define_metric(self, data, metric_name):
+        if "loss" in str.lower(metric_name):
+            wandb.define_metric(f"{data}-{metric_name}", summary="min")
+        elif str.lower(metric_name) in MINIMIZE_METRICS:
+            wandb.define_metric(f"{data}-{metric_name}", summary="min")
+        elif str.lower(metric_name) in MAXIMIZE_METRICS:
+            wandb.define_metric(f"{data}-{metric_name}", summary="max")
```

Comment:
include a maximize metric

Reply:
I didn't get this? Do you wanna have me write a test for maximize metric too?

Reply:
Doesnt have to be a separate test,  cant you add a metric that is named something that matches the maximize check to the existing test...    We dont have full coverage requirements but it seems you did all the work, might as well just add a few more lines to a test and get all this code covered.

Reply:
Added a metric but it's not meaningful from an ML perspective but I think that's not of a concern here. Thanks for the clarification.